File: finnssudokusite/sudoku/static/sudoku/js/solver/rules/rule_x.js
// === rule_x.js ===

import { Rule } from './rule.js';
import { EMPTY } from '../defs.js';
import { Candidates } from '../candidates.js';
import { Position } from '../position.js';

const X_TARGET = 10;

export class XRule extends Rule {
    /**
     * @param {Position} pos1
     * @param {Position} pos2
     */
    constructor(pos1, pos2) {
        super();
        this.pos1 = pos1;
        this.pos2 = pos2;
    }

    numberChanged(board, _changedCell) {
        return this.candidatesChanged(board);
    }

    candidatesChanged(board) {
        let changed = false;
        const c1 = board.getCell(this.pos1);
        const c2 = board.getCell(this.pos2);
        const a = c1.candidates;
        const b = c2.candidates;

        if (c1.value !== EMPTY && c2.value === EMPTY) {
            const need = X_TARGET - c1.value;
            for (let d = Candidates.MIN; d <= Candidates.MAX; d++) {
                if (b.test(d) && d !== need) {
                    b.disallow(d);
                    changed = true;
                }
            }
        } else if (c2.value !== EMPTY && c1.value === EMPTY) {
            const need = X_TARGET - c2.value;
            for (let d = Candidates.MIN; d <= Candidates.MAX; d++) {
                if (a.test(d) && d !== need) {
                    a.disallow(d);
                    changed = true;
                }
            }
        } else if (c1.value === EMPTY && c2.value === EMPTY) {
            for (let d = Candidates.MIN; d <= Candidates.MAX; d++) {
                const comp = X_TARGET - d;

                if (a.test(d) && (!b.test(comp) || comp < Candidates.MIN || comp > Candidates.MAX)) {
                    a.disallow(d);
                    changed = true;
                }

                if (b.test(d) && (!a.test(comp) || comp < Candidates.MIN || comp > Candidates.MAX)) {
                    b.disallow(d);
                    changed = true;
                }
            }
        }

        return changed;
    }

    checkPlausibility(board) {
        const c1 = board.getCell(this.pos1);
        const c2 = board.getCell(this.pos2);
        if (c1.value !== EMPTY && c2.value !== EMPTY) {
            return c1.value + c2.value === X_TARGET;
        }
        return true;
    }
}

File: rules/rule_x.cpp
// rule_x.cpp
#include "rule_x.h"
#include "../board.h"
#include "../candidates.h"

using namespace sudoku;

static constexpr int X_TARGET = 10;

XRule::XRule(Position p1, Position p2)
        : pos1_(p1), pos2_(p2)
{}

bool XRule::number_changed(Board& board, const Cell&) const {
    return candidates_changed(board);
}

bool XRule::candidates_changed(Board& board) const {
    bool changed = false;
    Cell* c1 = board.get_row(pos1_.row)[pos1_.col];
    Cell* c2 = board.get_row(pos2_.row)[pos2_.col];
    auto &a = c1->candidates;
    auto &b = c2->candidates;

    // one fixed → prune the other
    if (c1->value != EMPTY && c2->value == EMPTY) {
        int need = X_TARGET - c1->value;
        for (Candidate d = Candidates::MIN; d <= Candidates::MAX; ++d) {
            if (b.test(d) && d != need) {
                b.disallow(d);
                changed = true;
            }
        }
    }
    else if (c2->value != EMPTY && c1->value == EMPTY) {
        int need = X_TARGET - c2->value;
        for (Candidate d = Candidates::MIN; d <= Candidates::MAX; ++d) {
            if (a.test(d) && d != need) {
                a.disallow(d);
                changed = true;
            }
        }
    }
        // both unknown → prune impossible pairs
    else if (c1->value == EMPTY && c2->value == EMPTY) {
        for (Candidate d = Candidates::MIN; d <= Candidates::MAX; ++d) {
            if (a.test(d)) {
                int comp = X_TARGET - d;
                if (comp < Candidates::MIN || comp > Candidates::MAX || !b.test(comp)) {
                    a.disallow(d);
                    changed = true;
                }
            }
            if (b.test(d)) {
                int comp = X_TARGET - d;
                if (comp < Candidates::MIN || comp > Candidates::MAX || !a.test(comp)) {
                    b.disallow(d);
                    changed = true;
                }
            }
        }
    }

    return changed;
}

bool XRule::check_plausibility(const Board& board) const {
    Cell* c1 = board.get_row(pos1_.row)[pos1_.col];
    Cell* c2 = board.get_row(pos2_.row)[pos2_.col];
    if (c1->value != EMPTY && c2->value != EMPTY)
        return (c1->value + c2->value) == X_TARGET;
    return true;
}
